### PR TITLE
Add touch support

### LINKED ISSUE #138: Mobile support
It would be great if the tool supports mobile touch, so it can be used on devices such as tablets with a pencil.

---

Progress:

- [x] Responsive UI (#787)
- [x] touch support (#788)
- [x] make the touch areas for the resize handles larger (they shouldn’t appear visually larger though) (#790)
- [x] two-finger pan/zoom (#788)
- [x] fix press&hold on canvas triggering selection (maybe it’s not fixed? hard to tell) → fix by disabling user-select everywhere (#790)
- [x] edit text (#790)
- [x] delete objects (toolbar button?) (#788)
- [x] finalize line creation (#788)
  - [x] fix how existing `finalize` action deletes last line (#790)
- [x] refactor menu/edit toggles to be actions (#793)
- [x] make 🍔/🎨 buttons radios or otherwise show that they’re selected (#793)
- [ ] “reset zoom” button/indicator
- [x] undo/redo buttons (#793)
- [x] desktop pinch-zoom (Safari GestureEvents, Chrome/FF simulated ctrl+wheel)
- `apple-touch-icon` for adding to home screen ⏩ #820
- [ ] test `pointercancel`
- [ ] test two-finger gestures programmatically
- [ ] abort tap selection if a second `pointerdown` (i.e. for pan/zoom) occurs before the first pointer moves too far
- [ ] desktop pinch-zoom centered around cursor
- [ ] fix scrollbar offsets to account for toolbar (or resize canvas)
- re-add lock button somewhere ⏩ #819

Optional:
- [ ] multiple selection
- [ ] two-finger gestures while creating a line
- [ ] support releasing one finger from a two-finger gesture to switch back to select/edit mode (the reverse currently works)

### PR DESCRIPTION
**Pull request by j-f1**
More progress on #138.

- [x] touch support
- [ ] make the touch areas for the resize handles larger (they don’t have to appear visually larger though)
- [x] two-finger pan/zoom
- [x] fix press&hold on canvas triggering selection (maybe it’s not fixed? hard to tell)
- [ ] edit text
- [x] delete objects (toolbar button?)
- [x] finalize line creation
  - [ ] fix how existing `finalize` action deletes last line
- [ ] undo/redo

### CODE REVIEW

**Comment by j-f1** [comment]
It would be great if someone with an iPad could test this out because I don’t have one so I don’t know how well it would work. Specifically, holding space and dragging with one finger should work.

**Comment by nrgapple** [comment]
Is there an input for removing an element?

**Comment by j-f1** [comment]
Not yet @nrgapple

**Comment by j-f1** [comment]
I don’t understand why all the tests are failing

File: src/gesture.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { normalizeScroll } from "./scene/data";
+
+export function getCenter(pointers: readonly Pointer[]) {
+  return {
+    x: normalizeScroll(sum(pointers, p => p.x) / pointers.length),
+    y: normalizeScroll(sum(pointers, p => p.y) / pointers.length),
+  };
+}
+
+export function getDistance([a, b]: readonly Pointer[]) {
```
Comment (vjeux) [inline src/gesture.ts]: I learned that this is Math.hypot!

**Comment by vjeux** [comment]
Omg omg omg, just tried it on my phone and it’s working!!

3499B98F-E2E1-4707-A7DC-1E271878EEBC

One small thing, the hints location is off on mobile

**Comment by vjeux** [comment]
Probably not a bug introduced by this PR but because there’s no way to end a multi arrow selection, it let me try something unconventional:

- Click on the unlock icon to lock it
- Select arrow
- Click somewhere to start the arrow
- Click somewhere else to continue the arrow
- Click on the lock icon to unlock it
- The previous action resets the shape to selection and unselects the arrow
- Now you have both selection AND continue building the arrow

46EAED45-8339-47B9-B646-9B5D509572CF

Note: doesn’t need to be fixed in this PR, but we should keep track of it. I assume that when we’ll add some sort of done button (maybe next to the edit one), people won’t click on the lock anymore.

**Comment by vjeux** [comment]
For the editing, we’ll want to do something with the UI (don’t really know what yet) but now it’s almost unusable as it covers the entire scene.

53C8F72F-89E7-44C0-A86F-BF9D8F76D4F7

**Comment by vjeux** [comment]
For deletion, I can think of a trash icon next to the edit one. We’ll also likely want to put undo redo nearby. Maybe a way to copy paste as well

**Comment by vjeux** [comment]
Btw, I’m not opposed to disable the tests if you can’t find out why they are breaking. Mobile support is more important than the tests in the short term. Hopefully we’ll find the root cause and fix them afterwards.

**Comment by vjeux** [comment]
Seems like it’s green. Do you want to merge and keep iterating on subsequent PRs?

**Comment by j-f1** [comment]
Sure!

**Comment by dwelle** [comment]
Seems like https://github.com/excalidraw/excalidraw/pull/787 made the element properties menu appear even when no element is selected:

image

**Comment by j-f1** [comment]
Fixed @dwelle

**Comment by dwelle** [comment]
@j-f1 Cool. Tiny thing tho: you need to ensure the variable is always a boolean if you want to do a conditional rendering via `showSelectedShapeActions &&` :smile::

image

**Comment by lipis** [comment]
Now the elements property is not visible when inserting new elements..

**Pull request merged**